Fix frozen actors lacking tooltips if they have the cloak ability.

RoosterDragon opened the pull request:
Since bbf5970bc16a41ac5f7a7324c99dd0c3a917418d we update frozen actors only when required.

In 8339c6843ed122fb093b2f22e17a2c26e483b64d a regression was fixed where actors created in line of sight would be invisible.

Here, we fix a related regression where cloaked units that are revealed, and then frozen when you move out of line of sight would lack tooltips.

The fix centers around the setting of the Hidden flag. In the old code this used CanBeViewedByPlayer which checks for visibility modifiers and then uses the default visibility. The bug with this code is that when a visibility modifier was not hiding the actor, then we would report the default visibility state instead. However the frozen visibility state applies here which means if the frozen actor is visible, then we consider the actor to be hidden and therefore tooltips will not appear. In the fixed version we only consider the modifiers. This means a visibility modifier such as Cloak can hide the frozen actor tooltips. But otherwise we do not consider the frozen actor to be hidden. This prevents a frozen actor from hiding its own tooltips in some unintended circular logic. Hidden now becomes just a flag to indicate if the visibility modifiers are overriding things, as intended.

----

Fixes #20206. Regressed in #20156. See related regression #20179 and resolution in #20182

Ensure the following condition is met during the test scenarios:
- When frozen buildings are visible and you hover over one, they do not have healthbars and do have tooltips.

Test scenarios:
- When a building is created within line of sight, ensure it appears. Then move out of sight, and the frozen building appears.
- When a building is created out of line of sight, nothing is seen. Then move into line of sight and the building appears. Then move out of sight, and the frozen building appears.
- When using Explored Map = true, frozen buildings can be seen on the map when the game starts and can be targeted. e.g. oil derricks.
- When moving out of line of sight of a cloakable building that is cloaked, the frozen building does not appear.
- When moving out of line of sight of a cloakable building that is uncloaked, the frozen building appears.
- When enabling and then later disabling the `/visibility` cheat, frozen buildings appear for everything.

RoosterDragon committed:
Fix frozen actors lacking tooltips if they have the cloak ability.

Since bbf5970bc16a41ac5f7a7324c99dd0c3a917418d we update frozen actors only when required.

In 8339c6843ed122fb093b2f22e17a2c26e483b64d a regression was fixed where actors created in line of sight would be invisible.

Here, we fix a related regression where cloaked units that are revealed, and then frozen when you move out of line of sight would lack tooltips.

The fix centers around the setting of the Hidden flag. In the old code this used CanBeViewedByPlayer which checks for visibility modifiers and then uses the default visibility. The bug with this code is that when a visibility modifier was not hiding the actor, then we would report the default visibility state instead. However the frozen visibility state applies here which means if the frozen actor is visible, then we consider the actor to be hidden and therefore tooltips will not appear. In the fixed version we only consider the modifiers. This means a visibility modifier such as Cloak can hide the frozen actor tooltips. But otherwise we do not consider the frozen actor to be hidden. This prevents a frozen actor from hiding its own tooltips in some unintended circular logic. Hidden now becomes just a flag to indicate if the visibility modifiers are overriding things, as intended.

PunkPun reviewed (commented):
This fixes the debug bug of using /visibility twice unconditionally revealing all healthbars for the remainder of the game, but as a sideeffect now does not leave any frozen actors once you lose visibility

Activating GPS removes frozen actors if they are currently invisible (probably if sold or destroyed too). The frozen actor should stay, as it does normally when gps is running. Only the gps icon should be updated

RoosterDragon commented:
I checked the behaviour on e827e9952e45fb2d512a268fff112c044bd5fcb3 which is the commit before the change to frozen actors was made. What you describe appears to reflect the previous behaviour.

`/visibility` has an interesting behaviour where if you have it enabled when a building is built, you will get a frozen actor for it when you disable it again. If the building is built under fog, the cheat will reveal it but when you disable it you won't get a frozen actor. It's a bit odd but since it's the same before the regression inducing change I'm not concerned.

The GPS behaviour again seems to reflect bleed before the regression. I also think it makes sense. If the building cloaks, is sold or is destroyed then it should be removed because the GPS can tell it isn't there any more (or in the case of cloaking, has been fooled into as much).

PunkPun commented:
> I checked the behaviour on e827e9952e45fb2d512a268fff112c044bd5fcb3 which is the commit before the change to frozen actors was made. What you describe appears to reflect the previous behaviour.
>
> `/visibility` has an interesting behaviour where if you have it enabled when a building is built, you will get a frozen actor for it when you disable it again. If the building is built under fog, the cheat will reveal it but when you disable it you won't get a frozen actor. It's a bit odd but since it's the same before the regression inducing change I'm not concerned.

Did you use healthbars always on? Because as most units are undamaged you won't notice that anything changed, where as I see that this bug was finally fixed. Additionally on release frozen actors are created properly with /visibility, so in my eyes this IS a regression

> The GPS behaviour again seems to reflect bleed before the regression. I also think it makes sense. If the building cloaks, is sold or is destroyed then it should be removed because the GPS can tell it isn't there any more (or in the case of cloaking, has been fooled into as much).

To me it seems incorrect, because while GPS is active frozen actors work exactly the same as without having GPS active, the only weirdness that happens is at the moment of launch. But if it's not a regression fixing it here is unnecessary (opened an issue #20217)

RoosterDragon committed:
Notify of shroud state changes when using DevVisibility cheat.

When this cheat is used by notifying of shroud changes we invoke the usual logic that would occur if the visibility had been granted by units. Without this change any cached information about the visibility is not refreshed. Without this refresh actors with different visibility may not act correctly.

One aspect this improves is frozen actors. Using the visibility cheat will show up all actors on the map. If the cheat is then disabled than frozen actors will appear in their place. Prior to this change a frozen actor would fail to appear if the cheat had caused it to be revealed. Healthbars and selection boxes are also made consistent for similar reasons.

RoosterDragon commented:
> 
> Did you use healthbars always on? Because as most units are undamaged you won't notice that anything changed, where as I see that this bug was finally fixed. Additionally on release frozen actors are created properly with /visibility, so in my eyes this IS a regression
> 

I do not as a base, but even so the behaviour I described before is also present in the last release. Whether you get a frozen actor on disabling the cheat depends on whether the actor was visible before you enabled the cheat. Whether the healthbars disappear correctly is also affected by this condition.

That said, I have pushed a separate commit to improve the behaviour of using the `/visibility` cheat. If you enable and then later disable the cheat, frozen actors will appear for all actors (since the cheat revealed them all). With always show healthbars on, we also see healthbars and selection boxes disappearing on the frozen actors as expected.

Although that isn't technically fixing a regression as such, since it's making things consistent about how frozen actors behave it seems fair enough to lump in with the PR.

PunkPun commented on OpenRA.Game/Traits/Player/Shroud.cs:
```diff
@@ -101,20 +101,9 @@ public ShroudSource(SourceType type, PPos[] projectedCells)
 		// Per-cell cache of the resolved cell type (shroud/fog/visible)
 		readonly ProjectedCellLayer<ShroudCellType> resolvedType;
 
+		bool oldDisabled;
```
Perhaps this should use a different name. I find using `old` doesn't really fit here and just makes it confusing. Perhaps keep the custom `set` and add `disabledDirty = true` to it or instead add a comment / summary / remark

RoosterDragon replied:
Changed to keep custom getter/setter as suggested, and promoted `disabledChanged` to a field.

PunkPun reviewed (commented):
LGTM otherwise

PunkPun commented on OpenRA.Game/Traits/Player/Shroud.cs:
```diff
@@ -155,10 +157,11 @@ void INotifyCreated.Created(Actor self)
 				return;
 
 			anyCellTouched = false;
+			disabledChanged = false;
```
This line should be moved to the end of the function else the checks will fail

RoosterDragon replied:
Fixed.

PunkPun reviewed (dismissed):
LGTM

Mailaender commented on OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs:
```diff
@@ -81,13 +81,10 @@ void INotifyCreated.Created(Actor self)
-						// Needed so tooltips appear.
-						state.FrozenActor.Hidden = false;
-					}
+						UpdateFrozenActor(frozen, playerIndex);
+					frozen.RefreshHidden();
```
```suggestion
						UpdateFrozenActor(frozen, playerIndex);

					frozen.RefreshHidden();
```

Mailaender replied:
or
```suggestion
						{
							UpdateFrozenActor(frozen, playerIndex);
							frozen.RefreshHidden();
						}
```
?

RoosterDragon replied:
Tweaked to first suggestion.

PunkPun commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/48ae24a30636791adf2662a0843e0861856fd5e9...a24b6828b3113ed4451168aca5deb43a4bce2154)
